### PR TITLE
Implement allocation support for GPU sublocales

### PR DESCRIPTION
**Pull request by e-kayrakli**
This PR mainly implements special allocators for the GPU sublocale.

```chapel
on here.getChild(1) { // assume this is a gpu sublocale
  var a: [1..3] int;
  // ...
}
```

We want the array in the snippet above to be allocated on the GPU. In order to
do that, we need to make sure that both (1) array instance, and (2) its ddata is
allocated on the memory accessible by the GPU.

#### Background

Most of the heap allocations are managed by 5 functions implemented in
the LocaleModelHelpMem module. These are:

- `chpl_here_alloc`
- `chpl_here_calloc`
- `chpl_here_aligned_alloc`
- `chpl_here_realloc`
- `chpl_here_free`

These functions typically call into runtime functions that are named
`chpl_mem_*`.

However, the data allocations for the array are handled by separate but similar
functions implemented in `runtime/include/chpl-mem-array.h`. That's done to
register the array data with the comm layer for faster communication.

#### This PR:

- Implements `chpl_gpu_mem_*` functions that correspond to those 5 above.
- Implements `chpl_here_*` functions in the GPU locale model that calls into the
  appropriate runtime function depending on where we are executing etc.
- Adjusts the allocator and the deallocator in `chpl-mem-array` that's used by
  ddata allocations to use `chpl_gpu_*` as appropriate.

The end result is the snippet above ends up allocating the array's instance and
the ddata on the unified memory that's accesible from GPU and the CPU.

We eventually want to refactor this implementation. The idea is captured here:

https://github.com/Cray/chapel-private/issues/2406

#### Unified vs Device Memory

I initially started implementing this with device memory in mind. However,
there are some thorny areas there. If we use device memory, the array instance
will be allocated on the GPU memory, therefore, we'd need to initialize the
instance on the GPU. We aren't ready for that yet. There's also a question
whether we should actually do that sequential operation on the GPU or try to
move that instance from CPU to GPU after initialization for performance
purposes. More on this:

  https://github.com/Cray/chapel-private/issues/2296#issuecomment-894543245

#### More implementation details:

- Drops the number of arguments actual from the kernel launch primitives.
- Adjusts tests that invoke the kernel launch primitive.
- Adds a more proper `chpl_gpu_is_device_ptr` helper in the runtime.
- Lets `chpl_gpu_mem_memalign` give a "not implemented" error.
- Adds two new tests.

#### Test
- [x] `test/gpu/native` passes with the nightly GPU testing config
- [x] standard

### CODE REVIEW

File: runtime/include/chpl-mem-array.h [hunk 1]
```diff
@@ -54,6 +55,15 @@ static inline
 void* chpl_mem_array_alloc(size_t nmemb, size_t eltSize,
                            c_sublocid_t subloc, chpl_bool* callPostAlloc,
                            int32_t lineno, int32_t filename) {
+  void* p = NULL;
+#ifdef HAS_GPU_LOCALE
+  if (chpl_gpu_running_on_gpu_locale()) {
+    *callPostAlloc = false;
+    p = chpl_gpu_mem_alloc(nmemb*eltSize, CHPL_RT_MD_ARRAY_ELEMENTS,
+                           lineno, filename);
+  }
```
Comment (ronawho) [inline runtime/include/chpl-mem-array.h:64]: Note that as-implemented this skips our malloc hooks (which means we won't track memory allocations for GPU memory)

I'd also probably hoist the size computation from below above and use that here instead of doing the math again (not that it's hard math, but for uniformity in calls)

I think you'll also want to implement realloc for arrays too. I'm not sure that it's needed today since I think the module code is protected by a `LOCALE_MODEL=="flat"` check, but it'd be good to implement this code or put in assert so it's easier to catch if that support is extended in the future.

File: runtime/include/chpl-mem-array.h [hunk 2]
```diff
@@ -157,6 +168,12 @@ static inline
 void chpl_mem_array_free(void* p,
                          size_t nmemb, size_t eltSize,
                          int32_t lineno, int32_t filename) {
+#ifdef HAS_GPU_LOCALE
+  if (chpl_gpu_is_device_ptr(p)) {
```
Comment (ronawho) [inline runtime/include/chpl-mem-array.h:172]: Just curious -- but do you know how expensive this call is? I imagine it's quite slow.

File: compiler/codegen/cg-expr.cpp [hunk 3]
```diff
@@ -4706,26 +4706,33 @@ static GenRet codegenGPUKernelLaunch(CallExpr* call, bool is3d) {
...
+  int nNonKernelParamArgs = is3d ? 7:3;
+  int nKernelParamArgs = call->numActuals() - nNonKernelParamArgs;
+
   const char* fn = is3d ? "chpl_gpu_launch_kernel":"chpl_gpu_launch_kernel_flat";
 
   std::vector<GenRet> args;
   int i = 0;
   for_actuals(actual, call) {
-    if (i >= nNonKernelParamArgs) {
+    if (i++ > nNonKernelParamArgs) {
```
Comment (mppf) [inline compiler/codegen/cg-expr.cpp:4723]: I'm finding the logic of these conditionals and `i` hard to follow. Can you write the loop so that `i++` is the last statement in the `for_actuals` loop and not in a conditional?

File: runtime/include/chpl-mem-array.h (hunk 2 above)
Comment (mppf) [reply runtime/include/chpl-mem-array.h:172]: Since in a sublocale configuration we can use wide pointers to indicate which sublocale the pointer is referring to, I don't think we need to use this at all. I would imagine that we already have sublocale `int`s in the wide pointers when this PR and we could check the sublocale field of the wide pointer instead.

File: runtime/include/chpl-mem-array.h (hunk 1 above)
Comment (mppf) [reply runtime/include/chpl-mem-array.h:64]: I agree

File: runtime/src/chpl-gpu.c [hunk 4]
```diff
@@ -194,5 +248,90 @@ void chpl_gpu_launch_kernel_flat(const char* name, int grd_dim, int blk_dim,
...
+                            chpl_mem_descInt_t description,
+                            int32_t lineno, int32_t filename) {
+  CHPL_GPU_LOG("chpl_gpu_mem_memalign called. Size:%d\n", size);
+  chpl_internal_error("Not ready to allocate aligned memory on GPU, yet.");
+
+  // ENGIN: I don't know if it is possible to allocate memory with custom
+  // alignment on GPU. It looks like GPU's typically have a default alignment
+  // (512?) that cannot be changed. I don't think we'd need more than that
+  // today, and if we want, we can play some pointer games to return something
+  // with a smaller alignment here.
```
Comment (mppf) [inline runtime/src/chpl-gpu.c:323]: in the comment, do you mean "with a larger alignment" ?

File: modules/internal/localeModels/gpu/LocaleModel.chpl [hunk 5]
```diff
@@ -30,10 +30,167 @@
...
 
   public use LocaleModelHelpGPU;
-  public use LocaleModelHelpMem;
   require "-lcudart";
 
-  use IO;
+  use IO, SysCTypes, CPtr;
+
+  private inline
+  proc runningOnGPUSublocale(): bool {
```
Comment (mppf) [inline modules/internal/localeModels/gpu/LocaleModel.chpl:38]: Later on, when we call `chpl_gpu_init` implicitly, I'm not so sure we can check for a GPU context in this way.

Instead, I would expect it to only consider the sublocale.

Does this PR work if this function is just `chpl_task_getRequestedSubloc()>0` ?

**Review by mppf — APPROVED** [review]
> Adjusts the allocator and the deallocator in chpl-mem-array that's used by
ddata allocations to use chpl_gpu_* as appropriate.

Given that we are changing this in a locale-model-dependent way, should we move this function to the Chapel module locale model code as well? (If we decide to, I wouldn't do it in this PR. Perhaps @gbtitus knows a reason it should not have Chapel module code invoking it)

**Comment by gbtitus** [comment]
> Given that we are changing this in a locale-model-dependent way, should we move this function to the Chapel module locale model code as well? (If we decide to, I wouldn't do it in this PR. Perhaps @gbtitus knows a reason it should not have Chapel module code invoking it)

I don't, in fact I agree it would be preferable to do this logic in the Chapel module code that calls the `chpl_mem_array_*()` interface rather than in the runtime support as such. All of this seems very locale module-oriented to me, so maybe it even belongs not in the direct callers of `chpl_mem_array_*()` (which is the `_ddata_()` stuff in `ChapelBase`) but in callers of that. Another way to deal with this would be to extend the `chpl_mem_array` interface by adding GPU-oriented functions to it (so as to support tracking etc. if desired), but still move the decision as to which routine(s) to call up into the locale model-oriented code in the modules.

File: runtime/include/chpl-mem-array.h (hunk 1 above)
Comment (e-kayrakli) [reply runtime/include/chpl-mem-array.h:64]: > Note that as-implemented this skips our malloc hooks (which means we won't track memory allocations for GPU memory)

Yep.  I want to settle on https://github.com/Cray/chapel-private/issues/2404 before we make that connection, if that makes sense.

The rest, I'll fix.

File: runtime/include/chpl-mem-array.h (hunk 2 above)
Comment (e-kayrakli) [reply runtime/include/chpl-mem-array.h:172]: @ronawho -- My guess was that it would be fast as I was assuming a large chunk of the address space would be used for gpu-accessible memory, and that chunk would be easy to compute. But maybe it is more page-by-page or something, in which case it might be slow.

@mppf -- I think we can do that, but that check will probably have to happen at the module code. Which also is aligned with the general feedback that this whole logic should move to module code.

File: runtime/src/chpl-gpu.c (hunk 4 above)
Comment (e-kayrakli) [reply runtime/src/chpl-gpu.c:323]: Yes, indeed. Good catch.

File: modules/internal/localeModels/gpu/LocaleModel.chpl (hunk 5 above)
Comment (e-kayrakli) [reply modules/internal/localeModels/gpu/LocaleModel.chpl:38]: > Does this PR work if this function is just chpl_task_getRequestedSubloc()>0 ?

I agree that this'll be a problem when we initialize GPU context implicitly.

In my initial attempt where GPU sublocale allocations were made in the non-unified memory, doing what you suggest was problematic. When setting up the locales, you'd want the GPU sublocale instance to be on the CPU. That's why the logic currently checks if there's a GPU context (which you won't have during locale setup, resulting in all the allocations during setup be on the CPU).

With the use of unified memory, we may be able to get away with dropping that context check.

File: runtime/include/chpl-mem-array.h (hunk 2 above)
Comment (mppf) [reply runtime/include/chpl-mem-array.h:172]: FWIW we do have runtime calls into the generated module code, so you could call it from here even if this part wasn't yet migrated.

**Comment by e-kayrakli** [comment]
So, here's what I'm thinking:

- we add `chpl_here_array_alloc` etc functions in `LocaleModelHelpMem`, similar to `chpl_here_alloc` functions that are there today.
- Those functions call the `chpl_mem_array_*` interface as we have today.
- Then, we add `chpl_gpu_mem_array_*` functions to the GPU support.
- Finally, add `chpl_here_array_alloc` versions to the GPU locale model.

Basically make this "regular heap allocation" vs "array allocation" implementations similar. I think that can also have nice benefits for us where we can allocate an array instance on the unified memory but array's data in the device memory, for example.

@gbtitus -- This is a bit different than what you're suggesting, though. With what you suggest, we'd have the GPU-specific stuff in the `chpl_mem_array` interface. With what I am suggesting, that stuff will be in the `chpl-gpu` interface.

@gbtitus, @mppf, @ronawho -- Does this make sense? Any preference as to whether that happens as part of this PR, or a separate refactor effort?

**Comment by mppf** [comment]
Works for me. I don't have any strong preference for in this PR or not.

**Comment by e-kayrakli** [comment]
I tried to take a crack at it, bit it'll be a bigger refactor than I had imagined. The array allocators are called from ChapelBase, but ChapelBase is `use`d earlier than LocaleModel modules. So doing the first bullet caused some resolution errors.

The next thing to try would be is to separate `LocaleModelHelpMem` into separate helpers for each locale model, put those in separate paths next to where the corresponding `LocaleModel`s are, and `use` that before `use`ing `ChapelBase`. Note that `LocaleModelHelpMem` is just bunch of extern wrappers.

Doing that feels a bit much for this PR. But I'll wait to hear from Elliot and/or Greg about it.

File: modules/internal/localeModels/gpu/LocaleModel.chpl (hunk 5 above)
Comment (e-kayrakli) [reply modules/internal/localeModels/gpu/LocaleModel.chpl:38]: Ah.. that issue is still there. Because if you don't have a context, you cannot allocate unified memory, because you cannot use any cuda function. I'll look at this while working on making `chpl_gpu_init` calls implicit. (https://github.com/Cray/chapel-private/issues/2377)
